### PR TITLE
[GOBBLIN-1921] Properly handle reminder events

### PR DESCRIPTION
**Pull request by umustafi**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1921

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
Reminder flow trigger events were being improperly handled and interpreted as new events because they are triggered linger time after the original trigger where epsilon < linger and we use epsilon to determine event distinctness. With reminder events being considered distinct events, we were launching excess concurrent flows that were then being cancelled. Now we handle reminder events differently from normal event triggers to ensure they're properly evaluated. Because of db laundering, reminder events are easy to handle - if they're older than the currently worked upon event in the database they can be skipped and if they're equal to the current event in the database they are handled like normal. Reminder events should never be newer than the current event in the lease arbiter table because db laundering always results in increasing event times. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Updates existing unit tests 
TODO: adding tests to handle reminder events 

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/OrchestratorTest.java [hunk 1]
```diff
@@ -341,13 +341,13 @@ public void doNotRegisterMetricsAdhocFlows() throws Exception {
...
     FlowSpec adhocSpec = new FlowSpec(URI.create("flow0/group0"), "1", "", ConfigUtils.propertiesToConfig(flowProps) , flowProps, Optional.absent(), Optional.absent());
-    this.orchestrator.orchestrate(adhocSpec, flowProps, 0);
+    this.orchestrator.orchestrate(adhocSpec, flowProps, 0, false);
     String metricName = MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX, "group0", "flow0", ServiceMetricNames.COMPILED);
     Assert.assertNull(metricContext.getParent().get().getGauges().get(metricName));
 
     flowProps.setProperty("job.schedule", "0/2 * * * * ?");
     FlowSpec scheduledSpec = new FlowSpec(URI.create("flow0/group0"), "1", "", ConfigUtils.propertiesToConfig(flowProps) , flowProps, Optional.absent(), Optional.absent());
-    this.orchestrator.orchestrate(scheduledSpec, flowProps, 0);
+    this.orchestrator.orchestrate(scheduledSpec, flowProps, 0, false);
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/OrchestratorTest.java:350]: shall we add an additional case where the reminder is true?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java [hunk 2]
```diff
@@ -104,32 +116,50 @@ protected interface CheckedFunction<T, R> {
       + "VALUES(1, ?, ?) ON DUPLICATE KEY UPDATE epsilon=VALUES(epsilon), linger=VALUES(linger)";
   protected static final String WHERE_CLAUSE_TO_MATCH_KEY = "WHERE flow_group=? AND flow_name=? AND flow_action=?";
   protected static final String WHERE_CLAUSE_TO_MATCH_ROW = WHERE_CLAUSE_TO_MATCH_KEY
-      + " AND event_timestamp=? AND lease_acquisition_timestamp=?";
-  protected static final String SELECT_AFTER_INSERT_STATEMENT = "SELECT event_timestamp, lease_acquisition_timestamp, "
-    + "linger FROM %s, %s " + WHERE_CLAUSE_TO_MATCH_KEY;
+      + " AND event_timestamp=CONVERT_TZ(?, '+00:00', @@session.time_zone)"
+      + " AND lease_acquisition_timestamp=CONVERT_TZ(?, '+00:00', @@session.time_zone)";
```
Comment (phet) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:120]: don't we prefer to store and retrieve always as UTC?  if so, is the issue that `DEFAULT CURRENT_TIMESTAMP(3)` is what first sets us on to using the session TZ?  alternatively could we convert it to be UTC when it's initialized that way?  if we could, there wouldn't be a need for conversion here, would there be?

e.g. do we want everywhere want `UTC_TIMESTAMP(3)`, rather than `CURRENT_TIMESTAMP(3)`?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java [hunk 3]
```diff
@@ -186,12 +218,12 @@ private void initializeConstantsTable() throws IOException {
   }
 
   @Override
-  public LeaseAttemptStatus tryAcquireLease(DagActionStore.DagAction flowAction, long eventTimeMillis)
-      throws IOException {
-    log.info("Multi-active scheduler about to handle trigger event: [{}, triggerEventTimestamp: {}]", flowAction,
-        eventTimeMillis);
+  public LeaseAttemptStatus tryAcquireLease(DagActionStore.DagAction flowAction, long eventTimeMillis,
+      boolean isReminderEvent) throws IOException {
+    log.info("Multi-active scheduler about to handle trigger event: [{}, triggerEventTimestamp: {}]" +
```
Comment (phet) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:223]: same comment as in `Orchestrator` logging

**Review by phet — COMMENTED** [review]
nice work debugging and then turning around a quick rework of the impl!

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 4]
```diff
@@ -109,6 +109,7 @@ public class ConfigurationKeys {
   // Event time of flow action to orchestrate using the multi-active lease arbiter
   public static final String ORCHESTRATOR_TRIGGER_EVENT_TIME_MILLIS_KEY = "orchestratorTriggerEventTimeMillis";
   public static final String ORCHESTRATOR_TRIGGER_EVENT_TIME_NEVER_SET_VAL = "-1";
+  public static final String FLOW_IS_REMINDER_EVENT_KEY = "isReminderEvent";
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java:112]: Yes that makes sense to move in the future.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/FlowTriggerHandler.java [hunk 5]
```diff
@@ -278,6 +280,8 @@ public static JobDataMap updatePropsInJobDataMap(JobDataMap jobDataMap,
     // excess flows to be triggered by the reminder functionality.
     prevJobProps.setProperty(ConfigurationKeys.SCHEDULER_PRESERVED_CONSENSUS_EVENT_TIME_MILLIS_KEY,
         String.valueOf(leasedToAnotherStatus.getEventTimeMillis()));
+    // Use this boolean to indicate whether this is a reminder event
+    prevJobProps.setProperty(ConfigurationKeys.FLOW_IS_REMINDER_EVENT_KEY, String.valueOf(true));
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/FlowTriggerHandler.java:284]: Ah this is a typo, it should default to false nvm.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java (hunk 2 above)
Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:120]: From the documentation, the timestamp stored is in UTC but upon retrieval/or interpreting timestamps provided it automatically converts it to the current time zone (determined by the server's time). On our side we are utilizing all UTC timestamps so I convert it to session timezone as MySQL will interpret it as session timezone. I *think* using `UTC_TIMESTAMP` as the default won't change that so internally I am using session time zone with `CURRENT_TIMESTAMP` . I tried using `UTC_TIMESTAMP(3)` before and got a syntax error below. Debugging the timezone issues on the unit tests is what took the majority of time in this change and this configuration with the unit tests is what I could get to work and seem like it's using the same timezone everywhere. 

`Caused by: java.sql.SQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'UTC_TIMESTAMP(3), lease_acquisition_timestamp TIMESTAMP(3) NULL DEFAULT UTC_TIME' at line 1`

File: gobblin-runtime/src/test/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiterTest.java [hunk 6]
```diff
@@ -141,14 +142,14 @@ public void testAcquireLeaseSingleParticipant() throws Exception {
...
-        mysqlMultiActiveLeaseArbiter.tryAcquireLease(launchDagAction, eventTimeMillis);
+        mysqlMultiActiveLeaseArbiter.tryAcquireLease(launchDagAction, eventTimeMillis, false);
     Assert.assertTrue(fifthLaunchStatus instanceof MultiActiveLeaseArbiter.NoLongerLeasingStatus);
 
     // Tests CASE 6 of no longer leasing a distinct event in DB
     // Wait so this event is considered distinct and a new lease will be acquired
     Thread.sleep(EPSILON * 3/2);
     MultiActiveLeaseArbiter.LeaseAttemptStatus sixthLaunchStatus =
-        mysqlMultiActiveLeaseArbiter.tryAcquireLease(launchDagAction, eventTimeMillis);
+        mysqlMultiActiveLeaseArbiter.tryAcquireLease(launchDagAction, eventTimeMillis, false);
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-runtime/src/test/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiterTest.java:152]: Added all these test cases now (were still in progress before).

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/OrchestratorTest.java (hunk 1 above)
Comment (umustafi) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/OrchestratorTest.java:350]: I'm not really testing the orchestrate function with reminder true but rather the unit test would be testing the `tryAcquireLease` part. See tests I added.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java [hunk 7]
```diff
@@ -86,16 +88,26 @@ protected interface CheckedFunction<T, R> {
...
+      use epoch-millis in UTC locally
+    - Upon using any timestamps from local we convert the timezone from UTC to session
+    - We desire millisecond level precision and denote that with `(3)` for the TIMESTAMP types
+   */
   private static final String CREATE_LEASE_ARBITER_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS %s ("
       + "flow_group varchar(" + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + ") NOT NULL, flow_name varchar("
       + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + ") NOT NULL, " + " flow_action varchar(100) NOT NULL, "
-      + "event_timestamp TIMESTAMP, "
-      + "lease_acquisition_timestamp TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP, "
+      + "event_timestamp TIMESTAMP(3) DEFAULT CURRENT_TIMESTAMP(3), "
```
Comment (Will-Lo) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:109]: I thought we were going to make this an int?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java [hunk 8]
```diff
@@ -210,8 +242,23 @@ public LeaseAttemptStatus tryAcquireLease(DagActionStore.DagAction flowAction, l
...
       Timestamp dbCurrentTimestamp = getResult.get().getDbCurrentTimestamp();
 
+      // For reminder event, we can stop early if the reminder eventTimeMillis is older than the current event in the db
+      // because db laundering tells us that the currently worked on db event is newer and will have its own reminders
+      if (isReminderEvent) {
+        if (eventTimeMillis < dbEventTimestamp.getTime()) {
+          log.info("tryAcquireLease for [{}, eventTimestamp: {}] - dbEventTimeMillis: {} - A new event trigger is being"
+              + " worked on, so this older reminder will be dropped.", flowAction, eventTimeMillis, dbEventTimestamp);
+          return new NoLongerLeasingStatus();
+        } if (eventTimeMillis > dbEventTimestamp.getTime()) {
```
Comment (Will-Lo) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:252]: small nit, should put this if statement on a new line given it is not related to the old one

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java (hunk 7 above)
Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:109]: I converted everything to BIGINT and updated the tests but reverted back because I removed the database laundering after realizing that not having database laundering made it very hard to deal with `reminderEvents`.  We end up having to think about every possible case of what to do with reminder events and `withinEpsilon` and `afterLinger` or not while db laundering instead of using the `triggerTimeMillis` from the hosts may result in out of order timestamps if we have host drift. With db laundering, we use database timestamp and it becomes easier to do time conversions with timestamp type. There's a lot of back and forth that went into this

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java [hunk 9]
```diff
@@ -264,9 +265,9 @@ public void orchestrate(Spec spec, Properties jobProps, long triggerTimestampMil
         String flowExecutionId = flowMetadata.get(TimingEvent.FlowEventConstants.FLOW_EXECUTION_ID_FIELD);
         DagActionStore.DagAction flowAction =
             new DagActionStore.DagAction(flowGroup, flowName, flowExecutionId, DagActionStore.FlowActionType.LAUNCH);
-        flowTriggerHandler.get().handleTriggerEvent(jobProps, flowAction, triggerTimestampMillis);
-        _log.info("Multi-active scheduler finished handling trigger event: [{}, triggerEventTimestamp: {}]", flowAction,
-            triggerTimestampMillis);
+        flowTriggerHandler.get().handleTriggerEvent(jobProps, flowAction, triggerTimestampMillis, isReminderEvent);
+        _log.info("Multi-active scheduler finished handling trigger event: [{}, triggerEventTimestamp: {}]" +
+            "(is " + (isReminderEvent ? "reminder" : "original") + ")", flowAction, triggerTimestampMillis);
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java:270]: now that you're using the `{}` placeholders, it seems clearest to use throughout, as:
```
log.info("Multi-active ... event: [{}, triggerEventTimestamp: {}] (is {})", flowAction, ttMillis, isReminderEvent ? "reminder" : "original");
```
writing it that way makes spotting the lack of space before `"(is "` easier to see.

also, to entertain which we prefer:
```
event: [{1}, triggerEventTimestamp: {2}] (is {3})
{3} event: [{1}, triggerEventTimestamp: {2}]
event: [{1}, is: {3}, triggerEventTimestamp: {2}]
```
(numbered purely for illustration)

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java (hunk 2 above)
Comment (phet) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:120]: aren't we on mysql 8?  if so, how about this - https://dba.stackexchange.com/a/279140 ?
```
event_timestamp TIMESTAMP(3) NOT NULL DEFAULT (UTC_TIMESTAMP)
```
?

Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:120]: I tried variations of this and got either syntax errors or `Invalid default value for 'event_timestamp'. I'll instead set the session timezone to UTC before reading or inserting. 

```
event_timestamp TIMESTAMP(3) NOT NULL DEFAULT UTC_TIMESTAMP
event_timestamp TIMESTAMP(3) NOT NULL DEFAULT (UTC_TIMESTAMP)
event_timestamp TIMESTAMP(3) NOT NULL DEFAULT (UTC_TIMESTAMP(3))
event_timestamp TIMESTAMP(3) NOT NULL DEFAULT UTC_TIMESTAMP(3)
```

Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:120]: Set default to `NULL` and use `CURRENT_TIMESTAMP(3)` upon insertion or updates bc the timestamps are automatically interpreted in session timezone. 

Including for documentation purposes that I also tried setting the session timezone before making insertions/updates/table creation but had trouble executing _multi-query_ statements with a `SET` and `SELECT` because the latter needs to be executed with `preparedStatement.executeQuery` to return a `ResultSet` but multiple queries are only allowed with `preparedStatement.execute`. The JDBC connection url also needs to also `multi-query statements` as the default connection url does not allow it. The use of the former with the multi-statement query results in a SQL error 

`java.sql.SQLException: ResultSet is from UPDATE. No Data.`

**Review by Will-Lo — APPROVED** [review]

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java [hunk 10]
```diff
@@ -287,19 +290,19 @@ else if (leaseValidityStatus == 2) {
...
         int numRowsUpdated = attemptLeaseIfExistingRow(thisTableAcquireLeaseIfMatchingAllStatement, flowAction,
             true,true, dbEventTimestamp, dbLeaseAcquisitionTimestamp);
-        return evaluateStatusAfterLeaseAttempt(numRowsUpdated, flowAction, Optional.of(dbCurrentTimestamp));
+        return evaluateStatusAfterLeaseAttempt(numRowsUpdated, flowAction, Optional.of(dbCurrentTimestamp), isReminderEvent);
       } // No longer leasing this event
         if (isWithinEpsilon) {
-          log.debug("tryAcquireLease for [{}, eventTimestamp: {}] - CASE 5: Same event, no longer leasing event in db: "
-              + "terminate", flowAction, dbCurrentTimestamp.getTime());
+          log.debug("tryAcquireLease for [{}, is: {}, eventTimestamp: {}] - CASE 5: Same event, no longer leasing event"
+              + " in db", flowAction, isReminderEvent ? "reminder" : "original", dbCurrentTimestamp.getTime());
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/MysqlMultiActiveLeaseArbiter.java:301]: Yes I intentionally removed the word bc i noticed the other log statements only describe the case not the action that follows

**Review by phet — APPROVED** [review]
